File: JS9Ch/src/S_7_Follow_Up/Required_14/python_cpp/_558_Sliding_Window_Matrix_Maximum_Hard.cpp

class Solution {
public:
    /**
     * @param matrix an integer array of n * m matrix
     * @param k an integer
     * @return the maximum number
     */
    int maxSlidingWindow2(vector<vector<int>>& matrix, int k) {
        // Write your code here
        int n = matrix.size();
        if (n == 0 || n < k)
            return 0;
        int m = matrix[0].size();
        if (m == 0 || m < k)
            return 0;

        vector<vector<int>> sum(n + 1, vector<int>(m + 1));
        for (int i = 0; i <= n; ++i) sum[i][0] = 0;
        for (int i = 0; i <= m; ++i) sum[0][i] = 0;

        for (int i = 1; i <= n; ++i)
            for (int j = 1; j <= m; ++j)
                sum[i][j] = matrix[i - 1][j - 1] +
                            sum[i - 1][j] + sum[i][j - 1] - sum[i - 1][j - 1];

        int max_value = INT_MIN;
        for (int i = k; i <= n; ++i)
            for (int j = k; j <= m; ++j) {
                int value = sum[i][j] - sum[i - k][j] -
                            sum[i][j - k] + sum[i - k][j - k];

                if (value > max_value)
                    max_value = value;
            }
        return max_value;
    }
};

File: J9Ch/src/S_7_Follow_Up/Required_14/python_cpp/_558_Sliding_Window_Matrix_Maximum_Hard.py

class Solution:
    # @param {int[][]} matrix an integer array of n * m matrix
    # @param {int} k an integer
    # @return {int} the maximum number
    def maxSlidingWindow2(self, matrix, k):
        # Write your code here
        n = len(matrix)
        if n == 0 or n < k:
            return 0
        m = len(matrix[0])
        if m == 0 or m < k:
            return 0

        sum = [[0 for j in xrange(m + 1)] for i in xrange(n + 1)]
        for i in xrange(1, n + 1):
            sum[i][0] = 0;
        for i in xrange(1, m + 1):
            sum[0][i] = 0;

        for i in xrange(1, n + 1):
            for j in xrange(1, m + 1):
                sum[i][j] = matrix[i - 1][j - 1] + \
                            sum[i - 1][j] + sum[i][j - 1] - sum[i - 1][j - 1]

        max_value = -0x80000000
        for i in xrange(k, n + 1):
            for j in xrange(k, m + 1):
                value = sum[i][j] - sum[i - k][j] - \
                        sum[i][j - k] + sum[i - k][j - k]

                if value > max_value:
                    max_value = value

        return max_value